### PR TITLE
[Templates] Remove unnecessary requirements and intro flash attn

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

As @scottsun94 noted, the requirements that this PR removes are not needed to execute the llm finetuning template.
This PR also adds flash attention to the template and pins deepspeed.

### CODE REVIEW

**Comment by scottsun94** [comment]
I didn't run the lora part. But for 7B, 13B, and 70B full parameter fine-tuning, I can run the template code without these 2 dependencies. I didn't compare the performances, etc.

Someone should confirm whether they are needed:
- for lora
- for better performance
- etc.

**Comment by ArturNiederfahrenhorst** [comment]
Thanks. I'll m

File: doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt
Comment (kouroshHakha) [inline doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt:1]: we should pin deepspeed to the right version as well.

File: release/ray_release/byod/byod_finetune_llvms.sh [hunk 1]
```diff
@@ -16,9 +16,8 @@ pip3 install -U \
     pytorch-lightning==2.0.6 \
     "protobuf<3.21.0" \
     torchmetrics==1.0.3 \
-    lm_eval==0.3.0 \
-    tiktoken==0.1.2 \
     sentencepiece==0.1.99 \
     "urllib3<1.27" \
     git+https://github.com/huggingface/transformers.git@d0c1aeb \
     git+https://github.com/huggingface/peft.git@08368a1fba16de09756f067637ff326c71598fb3
+pip3 install -U flash-attn --no-build-isolation
```
Comment (aslonnie) [inline release/ray_release/byod/byod_finetune_llvms.sh:23]: can we pin a version please?

File: doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt [hunk 2]
```diff
@@ -8,7 +8,6 @@ wandb
 pytorch-lightning
 protobuf<3.21.0
 torchmetrics
-lm_eval==0.3.0
-tiktoken==0.1.2
 sentencepiece
-peft==0.7.0
\ No newline at end of file
+peft==0.7.0
+flash-attn --global-option="--no-build-isolation"
```
Comment (aslonnie) [inline doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt:13]: pin a version?

**Comment by ArturNiederfahrenhorst** [comment]
I requested review by @justinvyu , who'll review today.

File: doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-deepspeed
```
Comment (kouroshHakha) [inline doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt:1]: does this template work on 0.12.3? Last time we tried it used to have some problems above 0.10.3. Am I remembering a different context?

**Review by justinvyu — APPROVED** [review]
One question, but otherwise lgtm.

File: doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/Dockerfile [hunk 4]
```diff
@@ -10,6 +10,7 @@ RUN pip install --upgrade pip
 RUN pip install -U -r requirements.txt
 RUN pip install torch==2.1.1 --index-url https://download.pytorch.org/whl/cu121
 RUN pip uninstall bitsandbytes -y
+RUN pip install flash-attn==2.4.2 --global-option="--no-build-isolation"
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/Dockerfile:13]: Honestly, no clue. It's the recommended way of installing flash attention according to their github repo.

[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/Dockerfile:13]: Screenshot 2024-01-10 at 22 09 31

File: doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/requirements.txt:1]: 0.12.3 is the one that we install in the ray release byod. So I'd expect it to work.

**Comment by ArturNiederfahrenhorst** [comment]
Screenshot 2024-01-15 at 15 57 51

Release tests succeeded on AWS.

**Review by can-anyscale — APPROVED** [review]

**Pull request merged**